lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -579,33 +579,28 @@ def write(chunk)
             # Save the following fields if available, then clear them out to
             # allow for determining whether we should parse the log or message
             # field.
-            timestamp = record.delete('time')
-            severity = record.delete('severity')
-            trace = record.delete(@trace_key)
-            span_id = record.delete(@span_id_key)
-            insert_id = record.delete(@insert_id_key)
+            preserved_keys = [
+              'time', 'severity', @trace_key, @span_id_key, @insert_id_key
+            ]
 
             # If the log is json, we want to export it as a structured log
             # unless there is additional metadata that would be lost.
             record_json = nil
-            if record.length == 1
+            if (record.keys - preserved_keys).length == 1
               %w(log message msg).each do |field|
                 if record.key?(field)
                   record_json = parse_json_or_nil(record[field])
                 end
               end
             end
             unless record_json.nil?
+              # Propagate these if necessary. Note that we don't want to
+              # override these keys in the JSON we've just parsed.
+              preserved_keys.each { |key| record_json[key] ||= record[key] }
```

Comment:
Should this be
```
              preserved_keys.each { |key| record_json[key] ||= record[key] if record.key?(key) }
```
instead?

Otherwise the record might have a few `nil` fields when the `key` is not present in `record`, right?

```
2.4.3 :001 > record = {'a' => 'b'}
 => {"a"=>"b"}
2.4.3 :003 > record['c'] = nil
 => nil
2.4.3 :004 > record
 => {"a"=>"b", "c"=>nil}
2.4.3 :008 > record.to_json
 => "{\"a\":\"b\",\"c\":null}"
```

Reply:
Good call!